File: Project01/homework1/code.py
from scipy.io import loadmat
import scipy.fftpack
from matplotlib import pyplot as plt, image
import numpy as np
from matplotlib.pyplot import imread



trainingData = loadmat('TrainingSamplesDCT_8.mat')

TrainSampleDCT_FG = trainingData['TrainsampleDCT_FG']
TrainSampleDCT_BG = trainingData['TrainsampleDCT_BG']

BG_prior = len(TrainSampleDCT_BG)/(len(TrainSampleDCT_BG) + len(TrainSampleDCT_FG))
FG_prior = len(TrainSampleDCT_FG)/(len(TrainSampleDCT_BG) + len(TrainSampleDCT_FG))

BGX = np.zeros(64)
FGX = np.zeros(64)

for coffs in TrainSampleDCT_BG:
    x = coffs.argsort()[-2]
    BGX[x] += 1

for coffs in TrainSampleDCT_FG:
    x = coffs.argsort()[-2]
    FGX[x] += 1

plt.bar(range(64),BGX)
plt.bar(range(64),FGX)



plt.show()

BGX = BGX/len(TrainSampleDCT_BG)
FGX = FGX/len(TrainSampleDCT_FG)

img = image.imread("cheetah.bmp")/255
rgb_weights = [0.2989, 0.5870, 0.1140]

img = np.dot(img[...,:3], rgb_weights)/255

x_size,y_size = img.shape

zigzag = np.loadtxt("Zig-Zag Pattern.txt",dtype='i')
def dct2(a):
    return scipy.fftpack.dct( scipy.fftpack.dct( a, axis=0, norm='ortho' ), axis=1, norm='ortho' )

result = np.zeros(img.shape)

for i in range(x_size-7):
    for j in range(y_size-7):
        dct= dct2(img[i:(i+8),j:(j+8)])
        dct = dct.flatten()
        x = dct.argsort()[-2]
        if BGX[x]*BG_prior < FGX[x]*FG_prior:
            result[i][j] = 1
result = result
plt.imshow(result)

cheetah_mask = image.imread('cheetah_mask.bmp')
print((cheetah_mask!=result).sum()/(x_size*y_size))

File: Project01/homework1/code.m
load('TrainingSamplesDCT_8.mat');

back_data_size = size(TrainsampleDCT_BG);
fore_data_size = size(TrainsampleDCT_FG);
P_Y_back = back_data_size(1)/(back_data_size(1) + fore_data_size(1));
P_Y_fore = fore_data_size(1)/(back_data_size(1) + fore_data_size(1));

[~,training_feature_back] = max(TrainsampleDCT_BG(:,2:64),[],2);
[~,training_feature_fore] = max(TrainsampleDCT_FG(:,2:64),[],2);

fore_histogram = histogram(training_feature_fore,0:64);
hold on
back_histogram = histogram(training_feature_back,0:64);

P_X_back = back_histogram.Values / back_data_size(1);
P_X_fore = fore_histogram.Values / fore_data_size(1);

[img, map] = imread('cheetah.bmp');

img2 = im2double(img);

[img_x_size,img_y_size] = size(img2);
zigzag = readmatrix('Zig-Zag Pattern.txt');
cheetah_map = zeros(img_x_size,img_y_size);
for i = 1:img_x_size-7
    for j = 1:img_y_size-7
        block8x8 = img2(i:i+7,j:j+7);
        dct = dct2(block8x8);
        dct = abs(dct);
        [ii,jj] = find(dct == max(max(dct)));
        dct(ii,jj) = -Inf;
        [pos_x,pos_y] = find(dct == max(max(dct)));
        Feature = zigzag(pos_x,pos_y);
        if (P_X_fore(Feature+1)/P_X_back(Feature+1)) > (P_Y_back/P_Y_fore)
            cheetah_map(i,j) = 1;
        end
    end
end

imshow(cheetah_map);


[A2,B2] = imread('cheetah_mask.bmp');
A2 = A2/255;
falseness = sum(sum(xor(A2, cheetah_map))) / (img_x_size*img_y_size);





